### DISCUSSION TITLE
issue: select in fieldarray with dynamic option list gets confused on item removal

### DISCUSSION THREAD

**Discussion by krausest** [react-hook-form/react-hook-form#9730] [Q&A] [answered]
### Version Number

7.4.1

### Codesandbox/Expo snack

https://codesandbox.io/s/clever-morse-55pkuh?file=/src/App.tsx

### Steps to reproduce

1. Go to https://codesandbox.io/s/clever-morse-55pkuh?file=/src/App.tsx (based on V7 TS Template)
2. Preparation: Notice that the second row has 'B' selected, third row has 'C' selected. The list of options is computed dynamically according to the rule that it must not show values that are selected in the other select boxes.
3. Click on delete button next to the select showing 'B'
4. The row with 'B' is removed, but the value of the 'C' row now shows incorrectly has 'D' selected (fun fact: Safari shows 'B'). The state printed next to the select box shows that this is out of sync with the form state and it should still show 'C'.

There's essential the same code with the latest react-hook-form version 7.41.3 and a pure react implementation below it on: https://codesandbox.io/s/loving-jerry-ly185q?file=/src/App.tsx:2114-2118
This version has the same issue.

### Expected behaviour

In step 4 the second line should show 'C'.

When I convert the select to a controlled object (by simply setting its value) it works as expected, but I'd say this shouldn't be necessary.
When the list of options is static (by removing the filter operation in line 46) the selection of the field array shows the correct value. 

### What browsers are you seeing the problem on?

Firefox, Chrome, Safari

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by bluebill1049** [comment]
I don't think this is an issue related to react hook form.  since your input is uncontrolled:

```tsx
<option key={c.value} value={c.value}>
  {c.label}
</option>
 ```

Your key is going to get duplicated, I would consider building a controlled component for your usage instead especially since there is runtime on each row update with your select options.

**Comment by leapful** [comment, marked as answer]
@krausest @bluebill1049 

The problem is because React needs to perform mutation nodes (add or remove node) for each `useEffect` hook (used by `useWatch` hook) but in case removing `selections.[1].selected` (selected value 'B'), the sibling node `selections.[2].selected` (selected value 'C') is being re-used for next re-rendering.

1. By default, we have:

```
selections.[0].selected 'A'
selections.[1].selected 'B' 
selections.[2].selected 'C'
```

2. `useFieldArray` hook is executed (`remove` method) then React performs remove node action

```
selections.[0].selected 'A'
selections.[2].selected 'C'
```

3. `useWatch` hook is executed with old data of 'selections' where `selections.[0].selected`, `selections.[1].selected`, `selections.[2].selected` are still available

```
selections.[0].selected 'A'
selections.[2].selected 'C' with new options with some values have same key (like 'D') related to `selections.[1].selected` are re-rendered
```

4. React continues to perform update node action to alter `selections.[2].selected` becomes `selections.[1].selected ` with new data of `fields` from `useFieldArray` hook which is now having only 2 items ('A', 'C')

```
selections.[0].selected 'A'
selections.[1].selected 'C'
```

5. Since the altering of step 3 already pollute the key to determine selected value, the rendered result will be 'D' or 'B' (depends on React computation which is also following each browser API implementation) instead of 'C'

In order to make it work, please use attribute `selected` for HTML element `option` to help React determine correct selected value.

Working demo: https://codesandbox.io/s/winter-pond-5gh8tf?file=/src/App.tsx

**Reply by bluebill1049** [reply]
thanks, @leapful for the detailed explanation, I think if this is actually a good use case for a controlled component as well, the entire component state can be just isolated on its own and let react manage the state transition.

**Comment by krausest** [comment]
Thanks for your explanation. I tried to understand it and found something that (at least) I found interesting.
If I don't use field.id as the key, but an ID that is kept in the form state (i.e.<selection `key={data[rowIndex].id}` ...>) it works fine, but only since rhf 7.10:
https://codesandbox.io/s/musing-engelbart-vjip0o?file=/src/App.tsx
